From a8231ed7cfebbc18267c68220333dcdff432a7b4 Mon Sep 17 00:00:00 2001
From: chdugw_auto <jkulkarni@maxlinear.com>
Date: Wed, 1 Sep 2021 15:56:08 +0300
Subject: [PATCH] Merge pull request #3511 in SW_UGW/linux from
 bugfix/DRV_DPM_SW-527-GSW_CTP_PortAssignmentFree-fix to master

* commit '5e9dd0e9d24492da652288f478c90a4068b55fc6':
  DRV_DPM_SW-527 GSW_CTP_PortAssignmentFree fix reference counter issue
---
 drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
index d1b0b803f99a..cc481d2e8ae9 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
@@ -18846,6 +18846,7 @@ GSW_return_t GSW_CTP_PortAssignmentFree(void *cdev, GSW_CTP_portAssignment_t *pa
 {
 	ethsw_api_dev_t *gswdev = GSW_PDATA_GET(cdev);
 	u32 i, ret;
+	GSW_CTP_portConfig_t CtpConfig = {0};
 
 	if (gswdev == NULL) {
 		pr_err("%s:%s:%d", __FILE__, __func__, __LINE__);
@@ -18864,6 +18865,17 @@ GSW_return_t GSW_CTP_PortAssignmentFree(void *cdev, GSW_CTP_portAssignment_t *pa
 		goto UNLOCK_AND_RETURN;
 	}
 
+	CtpConfig.nLogicalPortId = parm->nLogicalPortId;
+	for (i = parm->nFirstCtpPortId; i < (parm->nFirstCtpPortId + parm->nNumberOfCtpPort); i++) {
+		ret = GSW_CtpPortConfigReset(cdev, &CtpConfig);
+		if (ret == GSW_statusErr) {
+			pr_err("GSW_CtpPortConfigSet returns ERROR\n");
+			ret = GSW_statusErr;
+			goto UNLOCK_AND_RETURN;
+		}
+		CtpConfig.nSubIfIdGroup++;
+	}
+
 	/*Set Index not in-Use to the CTPs assoicated with this logical port*/
 	for (i = parm->nFirstCtpPortId; i < (parm->nFirstCtpPortId + parm->nNumberOfCtpPort); i++)
 		gswdev->ctpportconfig_idx[i].IndexInUse = 0;
